### PR TITLE
RELATED: ONE-4838 - Rework how pivot table works with columns

### PR DESCRIPTION
**Pull request by lupko**
Code in this PR delivers full rework of how Pivot Table works with table columns.

Before: a lot of mojo with col IDs that has attribute, attribute element and measure IDs (extracted from bear-specific URI) in it. Parsing of IDs and getting essential information out of them and then matching against result descriptors and headers was all around the table.

Also before: ag-grid's ColDef structure was polluted with metadata essential for our implementation. This was leading to runtime warnings issued from ag-grid.

After: 
- Introduce TableDescriptor. 
- This is constructed from DVF's data access slices and series descriptors. The table descriptor contains well-defined column descriptors for both slicing (row) and data (measure) descriptors. The ID-s are now meaningless and generated in sequential order. 
- From these descriptors specific to our impl, code creates ag-grid's ColDefs which no longer contain any of our information. An instance of TableDescriptor is connected to pivot table lifecycle same as the initial data view and the rest of non-react state.
- All code that needs to work with table structure specific information does so by working with TableDescriptor it gets through params
- All ag-grid ColDefs and ColGroupDefs now have colId and groupId respectivelly and these always match the IDs of the table column descriptors. Table Descriptor has everything code may need.
- Reorganized the pivot table code base into multiple smaller 'boxes' focused on different tasks (data, structure, resizing)
- Deleted a lot of really messed up code, replaced it with more sane code with additional code documentation and a lot of invariant checking to kill bad usage fast
- Improved typing and type safety (although we are not fully out of the woods yet)

Impact:
- All over the table code. However most impacted were:
  -  getRowTotals (where code had to determine which measure columns have activated totals and which not)
  -  drilling - code to obtain drill intersection completely reworked (simplified) because once it gets hold of the col descriptor it has everything needed in it
  - sorting - simplified how sort items are created, as soon as code gets colId it can get full column descriptor and then creating sort item is trivial
  - resizing - simplified how column width items are creates (same story as sorting). Most of the changes in resizing were about passing tableDescriptor around and then getting essential info from it. Reworked most of resizing unit tests as they were doing some crazy-heavy mocking

Known bugs at the moment:
-  Measure-less tables do not work yet. The TableDescriptor factory cannot understand them (because dataAccess infra cannot understand them)

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2269/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3892/ : FAILURE in 5m 15s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2270/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3893/ : FAILURE in 5m 17s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2271/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3894/ : FAILURE in 7m 11s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2272/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3895/ : SUCCESS in 11m 53s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2273/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3896/ : SUCCESS in 11m 38s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2274/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3897/ : SUCCESS in 11m 43s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2281/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3904/ : SUCCESS in 19m 05s

**Comment by ivan-nejezchleb** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2285/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3907/ : FAILURE in 14m 09s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2286/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/644/ : UNSTABLE in 20m 20s

File: libs/sdk-ui-pivot/src/impl/data/dataSourceUtils.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import { GridApi } from "@ag-grid-community/all-modules";
+import ApiWrapper from "../base/agGridApiWrapper";
+import isEqual from "lodash/isEqual";
+import { IGridTotalsRow } from "./resultTypes";
+import { ROW_SUBTOTAL, ROW_TOTAL } from "../base/constants";
+import { IDimension } from "@gooddata/sdk-model";
+
+export const areTotalsChanged = (gridApi: GridApi | undefined, newTotals: IGridTotalsRow[]): boolean => {
+    const currentTotalsCount = gridApi?.getPinnedBottomRowCount() ?? 0;
+    const newTotalsCount = newTotals === null ? 0 : newTotals.length;
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/data/dataSourceUtils.ts:11]: What if `newTotals === undefined`? Would `newTotals?.length ?? 0` be safer?

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+ *
+ * The important thing to remember is that ColDefs and ColGroupDefs have same colId/groupId as their column descriptor
+ * counterparts.
+ *
+ *
+ * @alpha
+ */
+export class TableDescriptor {
+    public readonly zippedSliceCols: Array<[SliceCol, ColDef]> = [];
+    public readonly zippedLeaves: Array<[DataColLeaf, ColDef]> = [];
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts:57]: Can you please add some docs for the zipped collections?

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+
+    /**
+     * Tests whether the column with the provided id is the first (e.g. left-most) column in the table. Table always starts with
+     * slicing columns followed by data columns.
+     *
+     * @param c - column id, Column, ColDef or ColGroupDef from ag-grid
+     */
+    public isFirstCol(c: string | Column | ColDef | ColGroupDef): boolean {
+        invariant(c, "id of column or Column, ColDef or ColGroupDef must be provided");
+        const id = typeof c !== "string" ? agColId(c) : c;
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts:175]: Nitpick: this serialization/sanitization is repeated several times in the class. Woudl a utility function be useful even outside of it? Or maybe just make `agColId` take strings as well?

File: libs/sdk-ui-pivot/styles/scss/pivotTable.scss [hunk 4]
```diff
@@ -273,7 +273,6 @@ $table-menu-hover: var(
 
         .gd-cell-show-hidden {
             color: inherit;
-            font-weight: normal;
```
Comment (no23reason) [inline libs/sdk-ui-pivot/styles/scss/pivotTable.scss:276]: Why was this removed?

File: libs/sdk-ui-pivot/src/impl/tests/stickyRowHandler.test.ts [hunk 5]
```diff
@@ -141,8 +146,9 @@ describe("stickyRowHandler", () => {
         });
 
         describe("current scroll top belongs to different line than the last one", () => {
-            const fakeGetDisplayedRowAtIndex = (): any => ({ data: { a_123: "123" } });
-            const fakeGridApi = getFakeGridApi(fakeGetDisplayedRowAtIndex);
+            const fakeGetDisplayedRowAtIndex = (): any => ({ data: { r_0: "123" } });
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/tests/stickyRowHandler.test.ts:149]: This `r_0` etc. is leaking the logic of the prefixes from the factories. I wonder if we should have functions similar to `idRef` like `rowId(index: number)` and use it everywhere.

File: libs/sdk-ui-pivot/styles/scss/pivotTable.scss (hunk 4 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/styles/scss/pivotTable.scss:276]: It caused an issue with missing Bold font style (when drilling on elements is active), when last item in group of the same elements is rendered for smoother scroll out from the view 
https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts#L91
Found accidentally

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2290/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3911/ : SUCCESS in 11m 39s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2291/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3912/ : SUCCESS in 11m 40s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2293/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3914/ : SUCCESS in 11m 49s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2295/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3916/ : SUCCESS in 11m 49s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2296/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3917/ : FAILURE in 11m 47s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2297/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3918/ : FAILURE in 11m 46s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2298/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3920/ : FAILURE in 11m 42s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2300/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3922/ : FAILURE in 11m 43s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2302/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3924/ : FAILURE in 12m 05s

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx [hunk 6]
```diff
@@ -1133,10 +1085,14 @@ export class CorePivotTablePure extends React.Component<ICorePivotTableProps, IC
     };
 
     private async resetResizedColumn(column: Column) {
-        const id = getColumnIdentifier(column);
+        if (!this.tableDescriptor) {
+            return;
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/CorePivotTable.tsx:1089]: When can this happen?

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx [hunk 7]
```diff
@@ -1180,6 +1136,10 @@ export class CorePivotTablePure extends React.Component<ICorePivotTableProps, IC
     }
 
     private onColumnsManualReset = async (columns: Column[]) => {
+        if (!this.tableDescriptor) {
+            return;
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/CorePivotTable.tsx:1140]: When can this happen?

File: libs/sdk-ui-pivot/src/CorePivotTable.tsx [hunk 8]
```diff
@@ -1397,20 +1363,32 @@ export class CorePivotTablePure extends React.Component<ICorePivotTableProps, IC
                         cx(AG_NUMERIC_HEADER_CLASSNAME, "gd-measure-column-header"),
                     ),
                     // wrong params type from ag-grid, we need any
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/CorePivotTable.tsx:1365]: Nitpick: the comment is no longer valid.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2303/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3926/ : FAILURE in 11m 42s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2304/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3927/ : SUCCESS in 11m 54s

File: libs/sdk-ui-pivot/styles/scss/pivotTable.scss (hunk 4 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/styles/scss/pivotTable.scss:276]: Oh ok, thanks :)

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-ui-pivot/src/impl/tests/stickyRowHandler.test.ts (hunk 5 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/tests/stickyRowHandler.test.ts]: some of us are not ok with using the same "factories" for both generating the content and expected data in tests

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2307/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3930/ : FAILURE in 11m 56s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2308/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3931/ : FAILURE in 11m 59s

**Comment by ivan-nejezchleb** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2309/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3932/ : FAILURE in 14m 30s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2310/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/645/ : UNSTABLE in 19m 44s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2311/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3933/ : SUCCESS in 11m 51s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2312/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3934/ : SUCCESS in 11m 43s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2315/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3937/ : SUCCESS in 11m 46s

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts (hunk 2 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts]: done

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts (hunk 3 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptor.ts]: done

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2316/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3938/ : SUCCESS in 11m 34s

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+function getMinimalRowData(dv: DataViewFacade): DataValue[][] {
+    const data = dv.rawData().twoDimData();
+    const rowHeaders = dv.meta().allHeaders()[0];
+    const numberOfRowHeaderItems = (rowHeaders[0] || []).length;
+
+    return data.length > 0
+        ? data
+        : // if there are no measures only attributes
+          // create array of [null] of length equal to the number of row dimension headerItems
+          (Array(numberOfRowHeaderItems).fill([null]) as DataValue[][]);
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:31]: `Array.prototype.fill` is not supported by IE11 and we do not polyfill this. You can use lodash.fill instead.

File: libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import { DataValue } from "@gooddata/sdk-backend-spi";
+import { getMappingHeaderUri } from "@gooddata/sdk-ui";
+import { invariant } from "ts-invariant";
+import { IGridRow } from "../data/resultTypes";
+
+/**
+ * @deprecated this is linked to deprecated API
+ */
+function extractIdsFromAttributeElementUri(uri: string): (string | null)[] {
+    const [, attributeId, , attributeValueId = null] = uri.match(/obj\/([^/]*)(\/elements\?id=)?(.*)?$/)!;
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts:12]: Nitpick: the empty item in the middle of the array is quite error prone. We can fix this by using a non-capturing group in the regex

```ts
const previous = /obj\/([^/]*)(\/elements\?id=)?(.*)?$/
const improved = /obj\/([^/]*)(?:\/elements\?id=)?(.*)?$/
```

Then you would be able to remove the "empty" item in the destructuring (and _marginally_ improve performance).

File: libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        invariant(drillItemUri);
+
+        const maybeId = drillItemUri.startsWith("/gdc")
+            ? extractIdsFromAttributeElementUri(drillItemUri)[1]
+            : null;
+
+        result.push({
+            // Note: this is related to `id` deprecation. The whole `id` thing does not make sense. Code should
+            // send the entire URI (== PK of the element) so that the code is backend-agnostic. Doing the check
+            // here so that for bear, we follow this crap and for other backends code just adds the entire uri.
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts:73]: Not to be pedantic, but we should probably avoid saying crap in a public repo (if you remember the horse-related story)...

File: libs/sdk-ui-pivot/src/impl/resizing/tests/updateColDefs.test.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+import { TableDescriptor } from "../../structure/tableDescriptor";
+import {
+    testStore,
+    TwoMeasuresWithRowAttribute,
+    TwoMeasuresWithTwoRowAndTwoColumnAttributes,
+} from "./columnSizing.fixture";
+import { MANUALLY_SIZED_MAX_WIDTH, updateColumnDefinitionsWithWidths } from "../agGridColumnSizing";
+import { ReferenceData, ReferenceLdm } from "@gooddata/reference-workspace";
+
+// NOTE: the function under test mutate inputs (by design) so the tests
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/resizing/tests/updateColDefs.test.ts:20]: Probably missing the end of the sentence

File: libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts [hunk 13]
```diff
@@ -49,7 +50,16 @@ function shouldUpdate(
     return initialUpdate || differentRow || differentHorizontalBreakpoint;
 }
 
-export const updateStickyRowContentClasses = (
+const areDataDifferent = (previousData: any, currentData: any): boolean => {
+    return (
+        Object.keys(previousData).length !== Object.keys(currentData).length ||
+        Object.keys(previousData).some((dataItemKey: string) => {
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts:56]: Are we ok if this returns true for data that is effectively the same, just not referentially identical? Or should we use deep equality?

File: libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts (hunk 10 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts:12]: this function is move & rename of existing function we had in agGridUtils.ts/getIdsFromUri ... I just took what we have there and hidden it here without studying. if you really insist i will change, however I would prefer not to touch these regexp hieroglyphs :)

File: libs/sdk-ui-pivot/src/impl/structure/colLocatorMatching.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+                return undefined;
+            }
+
+            if (colMeasureLocalId(col) === locators[0].measureLocatorItem.measureIdentifier) {
+                found = col;
+            }
+        }
+
+        if (found) {
```
Comment (no23reason) [inline libs/sdk-ui-pivot/src/impl/structure/colLocatorMatching.ts:95]: Can we return the `found` items immediately? Or is there a way the `found` is re-assigned during one pass of the loop (I don't see one at first glance)?

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+            })
+            .map((total) => total.measureIdentifier);
+
+        totalsPerLeafColumn.forEach((value, idx) => {
+            const [leafDescriptor] = leafColumns[idx];
+
+            // if code bombs here then there must be something wrong in the table / datasource code because
+            // totals cannot (by definition) appear in a table that does not have measures - yet here we are
+            // processing totals
+            invariant(isDataColLeaf(leafDescriptor));
```
Comment (kandl) [inline libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:172]: Nitpick: a lot of invariants have a missing message (which in my opinion is a better development experience than reading the error description in the comment)

**Comment by no23reason** [comment]
Do we still need the `_old_agGridMock.ts` file?

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts (hunk 15 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:172]: You could argue thou, that those messages bloat the bundle (for cases that will hopefully not happen) while comments do not :-/

File: libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts (hunk 10 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts:12]: I understand, I do not insist, I just like regexes, I guess :D

File: libs/sdk-ui-pivot/src/impl/tests/stickyRowHandler.test.ts (hunk 5 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/src/impl/tests/stickyRowHandler.test.ts]: Ok, fair enough

File: libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts (hunk 13 above)
Comment (ivan-nejezchleb) [reply libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts:56]: data in this case are strings, so "return previousData[dataItemKey] !== currentData[dataItemKey]"  should be enough

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts (hunk 15 above)
Comment (kandl) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:172]: This should be probably solvable at the bundler level? Well, it's not really important, I just like when errors are as much descriptive as possible, because they can point you in the right direction even without checking the code & comments 😅

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+
+/**
+ * Convenience accessor of ag-grid column identifier. This can take either string itself or the
+ * different type of col flying around ag-grid API (ColDef, ColGroupDef or Column).
+ *
+ * For purposes of our table, column ID for ag-grid groups is the groupId.
+ *
+ * @param colOrId - column id (returned as is) or one of the ag-grid's column types
+ */
+export function agColId(colOrId: string | ColDef | ColGroupDef | Column): string {
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts:264]: I would move this accessor to the agGridUtils.ts where already `isColumn` predicate exists and where is imho better place for this

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+ * ----------
+ * val1  | 1
+ * val2  | 2
+ *
+ * There is one SliceCol - for row attribute A1. In this table, there will be one row for each
+ * element of attribute A1.
+ *
+ * TODO: rename to attributeCol?
+ */
+export type SliceCol = {
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts:24]: If the "slice" is the final name then change the "rowHeaderDescriptor" to something more slicier :)

File: libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts (hunk 13 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts:56]: @ivan-nejezchleb you are expert in the sticky row handler now, do you have some insight into this? :)

**Review by lupko — COMMENTED** [review]
.

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts (hunk 17 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts:24]: it's not final name. i will revisit the naming in subsequent PR, more names need to change.. even the DataCol* stuff

File: libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts (hunk 16 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/structure/tableDescriptorTypes.ts:264]: exactly my plan for a followup PR. want to to a bit more shaking there and possibly also rename some of the files.

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts (hunk 9 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:31]: interesting, this code was here for a while :)

Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:31]: i'm going to check it out though

Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts]: fixed

File: libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts (hunk 11 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/drilling/drilledRowFactory.ts]: fixed

File: libs/sdk-ui-pivot/src/impl/resizing/tests/updateColDefs.test.ts (hunk 12 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/resizing/tests/updateColDefs.test.ts]: fixed

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts (hunk 15 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:173]: i normally do this too for invariants that are on public API boundary. in this case, it's all in the internals and these invariants are there to catch bugs in the table code itself. if any of the invariant bombs, it does not mean the SDK user provided bad input.. it's that there is a bug in the table code itself. whoever ends up debugging can find in the comments possible reason why the invariant failed - where roughly did things go wrong and resulted in the bad invalid invariant.

that was the thinking behind.. these invariants are essentially `assert`s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-ui-pivot/src/impl/data/dataSourceUtils.ts (hunk 1 above)
Comment (lupko) [reply libs/sdk-ui-pivot/src/impl/data/dataSourceUtils.ts]: fixed

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2328/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3951/ : SUCCESS in 11m 44s

File: libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts (hunk 13 above)
Comment (no23reason) [reply libs/sdk-ui-pivot/src/impl/stickyRowHandler.ts:57]: ok, makes sense, thanks @ivan-nejezchleb, even though lodash.isEqual would be effectively the same here, right?

**Comment by lupko** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/2329/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/647/ : UNSTABLE in 14m 35s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result
 - No new bug found. Congratulation!

More detail, see in https://sonarqube-dev.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR993

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/713/ : SUCCESS in 26s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/976/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/3954/ : SUCCESS in 15m 49s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-el7/68/ : SUCCESS in 9m 51s (non-voting)
